src/lavinmq/federation/link.cr

```diff
@@ -232,7 +232,11 @@ module LavinMQ
             state(State::Running)
             unless @federated_q.immediate_delivery?
               @log.debug { "Waiting for consumers" }
-              @consumer_available.receive?
+              select
+              when @consumer_available.receive?
+              when timeout(1.second)
+                return if @upstream_connection.try &.closed?
+              end
```

Comment:
The timeout will cause a reconnect loop if no consumers are available. You probably need a loop around this select and break when consumer_available.

Reply:
Ah, yes, you're right 👍